### PR TITLE
mpfs: i2c: Add support for adaptive I2C bus frequency

### PR DESCRIPTION
**Pull request by jpaali**
Select the closest possible frequency which is smaller
than or equal to requested in I2C msg

### CODE REVIEW

File: arch/risc-v/src/mpfs/mpfs_i2c.c [hunk 1]
```diff
@@ -730,6 +739,73 @@ static int mpfs_i2c_reset(struct i2c_master_s *dev)
...
+        {
+          if (frequency >= mpfs_i2c_frequencies[i]
+              && mpfs_i2c_frequencies[i] > new_freq)
+            {
+              new_freq = mpfs_i2c_frequencies[i];
+              clock_div = i;
+            }
+        }
+
+      if (new_freq == 0)
```
[the remark this answers is not in the document]
Comment (jpaali) [reply arch/risc-v/src/mpfs/mpfs_i2c.c:821]: Yes, Our HW does not support 100kHz at the moment.

**Comment by jpaali** [comment]
Found a bug in init / reset functionality. It's fixed now and tested on ICICLE board.

File: arch/risc-v/src/mpfs/mpfs_i2c.c [hunk 2]
```diff
@@ -224,56 +246,65 @@ static void mpfs_restore_interrupts(irqstate_t primask)
...
 
-  modifyreg32(MPFS_I2C_CTRL,
-              MPFS_I2C_CTRL_CR2_MASK |
-              MPFS_I2C_CTRL_CR1_MASK |
-              MPFS_I2C_CTRL_CR0_MASK,
-              (((clock_speed >> 2u) & 0x01u) << MPFS_I2C_CTRL_CR2) |
-              (((clock_speed >> 1u) & 0x01u) << MPFS_I2C_CTRL_CR1) |
-              (((clock_speed & 0x01u) << MPFS_I2C_CTRL_CR0)));
+      priv->frequency = 0;
+      ret = mpfs_i2c_setfrequency(priv, I2C_DEFAULT_FREQUENCY);
```
Comment (jlaitine) [inline arch/risc-v/src/mpfs/mpfs_i2c.c:286]: Do you actually need this? The frequency is set at next transfer anyway?

Comment (jpaali) [reply arch/risc-v/src/mpfs/mpfs_i2c.c]: Yes, we don't need to call setfrequency() here. Actually CTRL register divider value is 0 after reset, which means 256 divider. Setting priv->frequency to mpfs_i2c_frequencies[0].

**Review by jlaitine — APPROVED** [review]

**Pull request merged**